---
title: "Assignment_3_answers"
author: "Jasmine Catague, Daisy Ta, Mihiravi Arachchige "
date: "21/05/2021"
output: html_document
---


# PART 1

# Question 1
### Present the RNA-seq count data in a table
Here is the downloaded file from github gene_expression2.tsv dataset containing the RNA-seq count data for two samples of interest. 

Firstly, we download the file by using the download.file command and then using the destfile command to label the file in our folder. 

To read the gene_expression2.tsv we can use read.table command selecting the "gene_expression.tsv" file and then using the command header=TRUE to read files with the labels in the first row. And then the command row.names is used to make the dataframe in to string vectors. 

``` {r gene_expression.tsv}

# Download the file gene_expression(.)tsv

download.file ("https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/gene_expression.tsv", 
  destfile= "gene_expression.tsv")
# Read in the file
gene_exp <- read.table("gene_expression.tsv", header=TRUE, row.names=1)
str(gene_exp)
# An example of trying to access a gene by gene name
gene_expression["ENSG00000227232", ]
# a table of values for the frst six genes.
gene_expression[1:6,]
```

# Question 2 
### Create a new column for the row means.

Create a new column for the means of the other columns using the rowMeans command. To create a new column use the command c(name). 
And then present it in a table with just the first six genes by using the command [1:6,] wherein [row,col]. 

```{r, new column for row means}

head(gene_exp)
# Calcualte the mean of the rows
mean_row<-rowMeans(gene_exp)
# Make a new mean column
gene_exp$mean<-c(mean_row)
# Show the first six genes
gene_exp[1:6,]
```


# Queston 3 
### List the 10 genes with the highest mean expression.

Firstly, we use the order command to sort the genes but this gives us the lowest to highest mean values of the genes. 

So then, we can use the tail command which selects the last rows of the dataset which corresponds to the top 10 highest gene expressions.
```{r, highest mean exp}
# Create sorted data frame from the ordered mean column
sorted_gene_exp<-gene_exp[order(gene_exp$mean),]
# Get the 10 highest mean values. 
top10_genes<-row.names(tail(sorted_gene_exp,10))
top10_genes
#show the result
  cat ("The top 10 genes with the highest mean values are", top10_genes)
```

# Question 4
### Determine the number of genes with a mean of <10

To determine the number of genes with a mean of more than 10 we can use the command sum (summary) and then select the mean value with <10. 
Then, using cat command we can separate character strings to print them properly. 

```{r, mean more than 10}
# Calculate the numbers of genes with a mean < 10
number_genes <-sum(gene_exp$mean<10)
number_genes
# Show the result
cat("The number of genes with a mean of more than 10 is", number_genes)


```
# Question 5 
### Make a histogram plot of the mean values 

To create a histogram we can use the hist command. 
```{r, histogram of means}

hist(gene_exp$mean, main= "The histogram of the mean values of the gene expressions.", xlab="Mean values", col="yellow")

```


# Question 6 
### Importing a .csv file 

Firstly, we can download the growth_data2.csv using the download.file command and then destfile command to save the file in the folder.

To read the file we can use the read.csv command and then use header=TRUE to read files with the labels in the first row. 
Then use the stringsAsFactors=FALSE command to re-encode strings.
```{r,growth dataset}
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/growth_data.csv", 
              destfile = "growth_data2.csv")

growth_data <-read.csv("growth_data2.csv", header= TRUE, stringsAsFactors =FALSE)

str(growth_data)
head(growth_data)
# Get the column names
colnames(growth_data)
```

# Question 7
### Calculate the mean and standard deviation of tree circumference at the start and end of the study at both sites.
Using the mean command we can calculate the mean of the tree circumference at the start of the study in 2004. Similarly, we can use the sd command to calculate the standard deviation of the tree circumference at the start of the study in 2004.

And then to calculate the tree circumference at the end of the study we can use the same commands mentioned above but selecting the 2019 datatset to calculate both mean and standard deviation at the end of the study. 
```{r, mean and sd }
# The mean and standard deviation of tree circumference at the start of the study in 2004 at the northeast.
northeast <- subset(growth_data,Site== "northeast")
head(northeast)
str(northeast)
mean_2004_nt <- mean(northeast$Circumf_2004_cm)
mean_2004_nt
cat("The mean of the tree circumference in the start of the study in 2004 at the northeast.", mean_2004_nt)

sd_2004_nt <-sd(northeast$Circumf_2004_cm)
sd_2004_nt
cat("The standard deviation of the tree circumference in the start of the study in 2004 at the northeast.", sd_2004_nt)

# The mean and standard deviation of tree circumference at the start of the study in 2004 at the southwest.
southwest<- subset(growth_data,Site== "southwest")
head(southwest)
str(southwest)

mean_2004_sw <- mean(southwest$Circumf_2004_cm)
mean_2004_sw
cat("The mean of the tree circumference in the start of the study in 2004 at the southwest.", mean_2004_sw)

sd_2004_sw <-sd(southwest$Circumf_2004_cm)
sd_2004_sw
cat("The standard deviation of the tree circumference in the start of the study in 2004", sd_2004_sw)


#The mean and standard deviation of tree circumference at the end of the study in 2019 at the northeast.
mean_2019_nt <- mean(northeast$Circumf_2019_cm)
mean_2019_nt
cat("The mean of the tree circumference in the end of the study in 2019 at the northeast.", mean_2019_nt)

sd_2019_nt <-sd(northeast$Circumf_2019_cm)
sd_2019_nt
cat("The standard deviation of the tree circumference in the end of the study in 2019 at the northeast.", sd_2019_nt)

#The mean and standard deviation of tree circumference at the end of the study in 2019 at the southwest.
mean_2019_sw <- mean(southwest$Circumf_2019_cm)
mean_2019_sw
cat("The mean of the tree circumference in the end of the study in 2019 at the southwest.", mean_2019_sw)

sd_2019_sw <-sd(southwest$Circumf_2019_cm)
sd_2019_sw
cat("The standard deviation of the tree circumference in the end of the study in 2019 at the southwest.", sd_2019_sw)


# Daisy's opinion
#The mean and standard deviation of tree circumference at the start (in 2004)
mean_2004 <- mean(growth_data$Circumf_2004_cm)
cat("Mean in 2004:", mean_2004)
sd_2004 <- sd(growth_data$Circumf_2004_cm)
cat("SD in 2004:", sd_2004)

#The mean and standard deviation of tree circumference at the end (in 2019)
mean_2019 <- mean(growth_data$Circumf_2019_cm)
cat("Mean in 2019:", mean_2019)
sd_2019 <- sd(growth_data$Circumf_2019_cm)
cat("SD in 2019:", sd_2019)
```

# Question 8 

### Make a boxplot of tree circumference at the start and end of the study at both sites
To make a boxplot, we used the boxplot command and then select the values we want to plot in this case we want the circumference of the study at both sides. 

We want the values from 2004 and 2014 northeast and southwest and then label them accordingly.
```{r, boxplot of tree circumference}


boxplot(northeast$Circumf_2004_cm, southwest$Circumf_2004_cm,northeast$Circumf_2014_cm, southwest$Circumf_2014_cm,
        main= "The tree circumference at the start and end of the study at both sites", 
        names=c("2004 at Northeast","2004 at Southwest",  "2014 at Northeast", "2014 at Southwest"), 
        xlab="Year",
        ylab="Tree Circumference (in cm)",
        col="orange")
        
 # Daisy's opinion
 # Get the tree circumference in 2004 and 20009
x_vals <- growth_data$Circumf_2004_cm
y_vals <- growth_data$Circumf_2019_cm
# Make a box plot
boxplot(x_vals, y_vals,
main = "Box plot of tree circumference",
names = c("2004", "2019"),
xlab = "Year",
ylab = "Tree circumference (cm)",
col = "orange"
)

```


# Question 9
### Calculate the mean growth over the past 10 years at each site.

To calculate the mean growth over the past 10 years at each site we can use the mean command selecting only the sites northeast and southwest using 2009 and 2019 datasets.

```{r, mean growth}


#The mean growth from the past 10 years.

#for northeast site
growth_2009_nt <- mean(northeast$Circumf_2009_cm)
head(growth_2009_nt)

growth_2019_nt <-mean(northeast$Circumf_2019_cm)
head(growth_2019_nt)
#difference is the mean growth in northeast site
mean_growth_nt= growth_2019_nt-growth_2009_nt
mean_growth_nt

#for southeast site
growth_2009_sw <- mean(southwest$Circumf_2009_cm)
head(growth_2009_sw)

growth_2019_sw <- mean(southwest$Circumf_2019_cm)
head(growth_2019_sw)
#difference is the mean growth in southeast site
mean_growth_sw= growth_2019_sw-growth_2009_sw
mean_growth_sw

```

# Question 10
### Use the t.test and wilcox.test functions to estimate the p-value that the 10 year growth is different at the two sites.

To use the t.test command and wilcox.test to get the p-value of the 10 year growth of the tree circumference from 2009-2019 and from northeast and southwest.
```{r, ttest}
# ten year growth in Northeast
northeast$tenyear_growth_nt<-northeast$Circumf_2019_cm-northeast$Circumf_2009_cm
head(northeast)
# ten year growth in Southwest
southwest$tenyear_growth_sw<-southwest$Circumf_2019_cm-southwest$Circumf_2009_cm
head(southwest)

t.test(northeast$tenyear_growth_nt,southwest$tenyear_growth_sw)

wilcox.test(northeast$tenyear_growth_nt,southwest$tenyear_growth_sw)

```

# PART 2
## Determine the limits of BLAST

# Question 1
To download the whole E.coli gene sequences we can use the download.file command. Then, we can decompress the file by using the gunzip command
```{r, Ecoli DNA seq}
# Loading the library
suppressPackageStartupMessages({
  library("seqinr")
  library("magrittr")
  library("kableExtra")
  library("R.utils")
  library("rBLAST")
  library("Biostrings")
  source("https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part2_files/mutblast_functions.R")
})
# Download the whole set of E. coli gene DNA sequences
download.file("ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/cds/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz",
              destfile = "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz")
# Use gunzip to decompress
R.utils::gunzip("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz", overwrite=TRUE)
# Create a BLAST database
makeblastdb("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa", dbtype="nucl", "-parse_seqids")
```
# Question2
We use download.file to download the sample fasta sequence. Then, we use seqinr::read.fasta to read them in. We use the command [] to choose our interest sequence (number 1). After that, we use getLength() and seqinr::GC() to determine the length and the proportion of GC bases respectively. Finally, we use cat() to show the results. 
```{r Q2Part2}
# Download the sample fasta sequences
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part2_files/sample.fa",
              destfile = "sample.fa")
# Read the sequences
SEQ <- seqinr::read.fasta("sample.fa")
str(SEQ)
# Select one sequence
my_seq <- SEQ[[1]]
# The length of my selected sequence
seqinr::getLength(my_seq)

# The proportion of GC bases
seqinr::GC(my_seq)
# Show the results
cat(" The length of my selected sequence is", seqinr::getLength(my_seq))
cat("The proportion of GC bases is", seqinr::GC(my_seq))
```
## Question 3
We use myblastn_tab to perform BLAST search. Then, the str command is used to see the structure. After that, we as as.character to identify what E.coli gene my sequence matches best. There is one sequence meet the demand because this gene does not share any matches with any other genes in the database.
```{r Q3Part2}
# Create BLAST databases and perform BLAST searches
myblastn_tab
res <- myblastn_tab(myseq= my_seq, db = "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa")
str(res)
#  Identify what E. coli gene my sequence matches best
top_hits <- as.character(res$sseqid[1:3])
top_hits
# Show a table of the top 3 hits
head(res, 3)[, c("qseqid", "sseqid", "pident", "evalue","bitscore")]
```
## Question 4
We use mutator() function to create mutated sequence with 100 point mutations and then compare with the original sequence by making a pairwise alignment by pairwiseAlignment from Biostrings library. Firstly, we read in my selected sequence. Then, we create a mutated copy with 100 substitutions. After that, we use DNAString() to convert to biostring. We use mmismatch() to get the number of mismatch. Finally, we show the result by using cat(). The number of mismatches between the original and mutated sequence is 80.
```{r Q4Part2}
mutator
# Read in
tophit <- seqinr::read.fasta("sample.fa")
tophit <- tophit[[1]]
str(tophit)

# Mutate
tophit_mut <- mutator(myseq = tophit, nmut = 100)

## Perform pairwise alignment to prove that the mutation has worked as expected
# Convert to biostring
tophit_f <- DNAString(c2s(tophit))
tophit_mut_f <- DNAString(c2s(tophit_mut))
aln <- Biostrings::pairwiseAlignment(tophit_f,tophit_mut_f)
pid(aln)

cat("The proportion of matched identity is", pid(aln)) 
# Get the number of mismatch
 nmismatch(aln)
# Show the result
 cat("The number of mismatches between the original and mutated sequence is", nmismatch(aln))
```
## Question 5
To determine the number and proportion of sites that need to be altered to prevent the BLAST search from matching the gene of origin we can 
```{r Q5Part2}

#We need to write the blast index first
makeblastdb(file="sample.fa", dbtype="nucl")

#From the previous data we know that we have 81 mismatches so we can test with 81 mismatches
tophit_mut<-mutator(myseq=tophit,nmut=81)

res<-myblastn_tab(myseq=tophit_mut,db="sample.fa")
res

!is.null(res)

cat("The result for any matches of the mutated sequence is",!is.null(res) )

#Now we increase the number of mutations to see whether we can still find matches.
tophit_mut<-mutator(myseq=tophit,nmut=200)

res<-myblastn_tab(myseq=tophit_mut,db="sample.fa")
res

!is.null(res)

cat("The result for any matches of the mutated sequence is",!is.null(res) )

rep_mut<-function(NMUT){
  tophit_mut<-mutator(myseq=tophit,nmut=NMUT)
  res<-myblastn_tab(myseq=tophit_mut,db="sample.fa")
  as.numeric(!is.null(res))
}

replicate(n=100, rep_mut(200))

#The sites that we have changed for our 650bp sequence.
n_mut_range<-c(0,30,50,80,110,140,170,200)

replicator_function<-function(NMUT){
  mean(replicate(n=100,rep_mut(NMUT)))
}

myresult<-sapply(n_mut_range, replicator_function)

myresult
```
# Question 6

Now we can construct the increasing proportion of mutated bases which reduces the ability for BLAST to match the gene origin.
```{r, chart}
#To create a chart we can make a scatterplot.

#The proportion of sites randomised from the total sequence length of 650bp
x<-c(0,0.04,0.07,0.12,0.16,0.21,0.26,0.31)

  #The proportion of successful BLASTs
y<-c(1.00, 0.99, 0.70, 0.45, 0.19, 0.04)

plot(x,y, 
     main="The effect of increasing number of random bases in BLAST", 
     xlab="proportion of sites randomised", 
     ylab="proportion of succesful BLASTs")

abline(lm(y~x),col="read")
```


