---
title: "Assignment_3_answers"
author: "Jasmine Catague, Daisy Ta, Mihiravi Arachchige "
date: "21/05/2021"
output: html_document
---


# PART 1: Importing files, data wrangling, plots and saving code on GitHub

## Question 1
Here is the downloaded file from github gene_expression2.tsv dataset containing the RNA-seq count data for two samples of interest. Firstly, we download the file by using the download.file command and then using the destfile command to label the file in our folder. To read the gene_expression2.tsv we use read.table command selecting the "gene_expression.tsv" file and then using the command header=TRUE to read files with the labels in the first row. After that, we use [] to access a gene by a gene name. Finally, we output the first six rows using [1:6,].

``` {r gene_expression.tsv}

# Download the file gene_expression(.)tsv

download.file ("https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/gene_expression.tsv", 
  destfile= "gene_expression.tsv")
# Read in the file
gene_exp <- read.table("gene_expression.tsv", header=TRUE, row.names=1)
str(gene_exp)
# An example of trying to access a gene by gene name
gene_expression["ENSG00000227232", ]
# a table of values for the first six genes.
gene_expression[1:6,]
```

## Question 2 
We create a new column for the means of the other columns by using the rowMeans command. Then, we create a new column use the command c(name). Finally, we present it in a table with just the first six genes by using the command [1:6,] wherein [row,col].
```{r, new column for row means}
# Calculate the mean of the rows
mean_row<-rowMeans(gene_exp)
# Make a new mean column
gene_exp$mean<-c(mean_row)
# Show the first six genes
gene_exp[1:6,]
```
## Question 3 
Firstly, we use the order command to sort the genes but this gives us the lowest to highest mean values of the genes. So then, we use the tail command which selects the last rows of the data set which corresponds to the top 10 highest gene expressions.
```{r, highest mean exp}
# Create sorted data frame from the ordered mean column
sorted_gene_exp<-gene_exp[order(gene_exp$mean),]
# Get the 10 highest mean values. 
top10_genes<-row.names(tail(sorted_gene_exp,10))
top10_genes
```

## Question 4
To determine the number of genes with a mean of more than 10 we use the command sum() and then select the mean value with <10. Then, using cat command we can separate character strings to print them properly. 

```{r, mean}
# Calculate the numbers of genes with a mean < 10
number_genes <-sum(gene_exp$mean<10)
number_genes
# Show the result
cat("The number of genes with a mean of < 10 is", number_genes)
```
## Question 5 
To create a histogram we use the hist command. 
```{r, histogram of means}
hist(gene_exp$mean, main= "The histogram of the mean values of the gene expressions", xlab="Mean values", col="yellow")
```
## Question 6 

Firstly, we download the growth_data2.csv by using the download.file command and then destfile command to save the file in the folder.To read the file we use the read.csv command and then use header=TRUE to read files with the labels in the first row. Then use the stringsAsFactors=FALSE command to re-encode strings. We use str and head command to make sure the data has been successful imported. Finally, we use colnames command to get the column names. 
```{r,growth dataset}
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/growth_data.csv", 
              destfile = "growth_data2.csv")

growth_data <-read.csv("growth_data2.csv", header= TRUE, stringsAsFactors =FALSE)
str(growth_data)
head(growth_data)
# Get the column names
colnames(growth_data)
```

## Question 7
We use the mean and sd command to calculate mean and standard deviation of tree circumference respectively in 2004 and 2019. Then, we use the cat command to show the results. Mean and sd of tree circumference in 2004 are 5.077 and 1.054462 respectively. Mean and sd of tree circumference in 2019 are 49.912 and 22.17979 respectively
```{r, mean and sd }
#The mean and standard deviation of tree circumference at the start (in 2004)
mean_2004 <- mean(growth_data$Circumf_2004_cm)
cat("Mean in 2004 is", mean_2004)
sd_2004 <- sd(growth_data$Circumf_2004_cm)
cat("SD in 2004 is", sd_2004)

#The mean and standard deviation of tree circumference at the end (in 2019)
mean_2019 <- mean(growth_data$Circumf_2019_cm)
cat("Mean in 2019 is", mean_2019)
sd_2019 <- sd(growth_data$Circumf_2019_cm)
cat("SD in 2019 is", sd_2019)
```

## Question 8 
Firstly, we get the tree circumference in 2004 and 2019 by using $ sign. Then, we use boxplot to make a box plot of tree circumference at the start and end of the study at both sites. The name of the box plot is "Box plot of tree circumference". 
```{r, boxplot of tree circumference}
# Get the tree circumference in 2004 and 20009
x_vals <- growth_data$Circumf_2004_cm
y_vals <- growth_data$Circumf_2019_cm
# Make a box plot
boxplot(x_vals, y_vals,
main = "Box plot of tree circumference",
names = c("2004", "2019"),
xlab = "Year",
ylab = "Tree circumference (cm)",
col = "orange"
)
```


## Question 9
Firstly, we get the tree circumference values of each site by using subset(). Then, we use mean command to calculate the mean values in 2009 and 2019 of each site before calculate. The mean growth over the past 10 years of Northeast and Southwest are 30.076 and 48.354 respectively.
```{r, mean growth}
# Get the tree circumference values of each site
northeast <- subset(growth_data, Site=="northeast")
southwest <- subset(growth_data, Site=="southwest")
# Calculate the mean values of each site over past 10 years
na_2019 <- mean(northeast$Circumf_2019_cm)
sw_2019 <- mean(southwest$Circumf_2019_cm)
na_2009 <- mean(northeast$Circumf_2009_cm)
sw_2009 <- mean(southwest$Circumf_2009_cm)
# Calculate the mean growth of each site
na_mean_growth <- na_2019 - na_2009
sw_mean_growth <- sw_2019 - sw_2009
cat("The mean growth over the past 10 years of Northeast is", na_mean_growth)
cat("The mean growth over the past 10 years of Southwest is", sw_mean_growth)
```

## Question 10
We get the growth values of tree circumference on each site by using $. Then, we use t.test() and wilcox.test() to run t.test and wilcox.test before getting the p-value. We use cat() command to show the result. The p-value of t.test is 1.712524e-06 and p-value of wilcox.test is 4.6264e-06.
```{r, ttest}
#get the growth values of each site
na_growth <- northeast$Circumf_2019_cm - northeast$Circumf_2009_cm
sw_growth <- southwest$Circumf_2019_cm - southwest$Circumf_2009_cm
#run t.test
t <- t.test(na_growth, sw_growth)
#get p-value of t.test
t_pvalue <- t$p.value
cat("p-value of t.test is", t_pvalue)
#run wilcox.test
wilcox <- wilcox.test(na_growth, sw_growth)
# get p-value of wilcox.test
wilcox_pvalue <- wilcox$p.value
cat("p-value of wilcox.test is", wilcox_pvalue)
```

# PART 2: Determine the limits of BLAST
```{r loading the library}
# Loading the library
suppressPackageStartupMessages({
  library("seqinr")
  library("magrittr")
  library("kableExtra")
  library("R.utils")
  library("rBLAST")
  library("Biostrings")
  source("https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part2_files/mutblast_functions.R")
})
```
## Question 1
We use download.file to download the whole set of E. coli gene DNA sequences and and use R.utils::gunzip to decompress. Then, we use makeblastdb() function to create a BLAST database with dbtype is nucleic. There are 4140 sequences which present in the E.coli set. 
```{r, Ecoli DNA seq}
# Download the whole set of E. coli gene DNA sequences
download.file("ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/cds/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz",
              destfile = "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz")
# Use gunzip to decompress
R.utils::gunzip("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz", overwrite=TRUE)
# Create a BLAST database
makeblastdb("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa", dbtype="nucl", "-parse_seqids")
```
# Question2
We use download.file to download the sample fasta sequence. Then, we use seqinr::read.fasta to read them in. We use the command [] to choose our interest sequence (number 1). After that, we use getLength() and seqinr::GC() to determine the length and the proportion of GC bases respectively. Finally, we use cat() to show the results. The length of my selected sequence is 615. The proportion of GC bases is 0.5560976
```{r Q2Part2}
# Download the sample fasta sequences
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part2_files/sample.fa",
              destfile = "sample.fa")
# Read the sequences
SEQ <- seqinr::read.fasta("sample.fa")
# Select one sequence
my_seq <- SEQ[[1]]
# The length of my selected sequence
length_seq <- seqinr::getLength(my_seq)
# The proportion of GC bases
seqinr::GC(my_seq)
# Show the results
cat(" The length of my selected sequence is", length_seq)
cat("The proportion of GC bases is", seqinr::GC(my_seq))
```
## Question 3
We use myblastn_tab to perform BLAST search. Then, the str command is used to see the structure. After that, we as as.character to identify what E.coli gene my sequence matches best. There is one sequence meet the demand because this gene does not share any matches with any other genes in the database.
```{r Q3Part2}
# Create BLAST databases and perform BLAST searches
myblastn_tab
res <- myblastn_tab(myseq= my_seq, db = "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa")
str(res)
#  Identify what E. coli gene my sequence matches best
top_hits <- as.character(res$sseqid[1:3])
top_hits
# Show a table of the top 3 hits
head(res, 3)[, c("qseqid", "sseqid", "pident", "evalue","bitscore")]
```
## Question 4
We use mutator() function to create mutated sequence with 100 point mutations and then compare with the original sequence by making a pairwise alignment by pairwiseAlignment from Biostrings library. Firstly, we read in my selected sequence. Then, we create a mutated copy with 100 substitutions. After that, we use DNAString() to convert to biostring. We use mmismatch() to get the number of mismatch. Finally, we show the result by using cat(). The number of mismatches between the original and mutated sequence is 80.
```{r Q4Part2}
# Read in
tophit <- seqinr::read.fasta("sample.fa")
tophit <- tophit[[1]]
str(tophit)
# Mutate
tophit_mut <- mutator(myseq = tophit, nmut = 100)
## Perform pairwise alignment to prove that the mutation has worked as expected
# Convert to biostring
tophit_f <- DNAString(c2s(tophit))
tophit_mut_f <- DNAString(c2s(tophit_mut))
aln <- Biostrings::pairwiseAlignment(tophit_f,tophit_mut_f)
pid(aln)
# Get the number of mismatch
 nmismatch(aln)
# Show the result
cat("The number of mismatches between the original and mutated sequence is", nmismatch(aln))
```
## Question 5
To determine the number and proportion of sites that need to be altered to prevent the BLAST search from matching the gene of origin we can 
```{r Q5Part2}

#We need to write the blast index first
makeblastdb(file="sample.fa", dbtype="nucl")

#From the previous data we know that we have 81 mismatches so we can test with 81 mismatches
tophit_mut<-mutator(myseq=tophit,nmut=81)

res<-myblastn_tab(myseq=tophit_mut,db="sample.fa")
res

!is.null(res)

cat("The result for any matches of the mutated sequence is",!is.null(res) )

#Now we increase the number of mutations to see whether we can still find matches.
tophit_mut<-mutator(myseq=tophit,nmut=200)

res<-myblastn_tab(myseq=tophit_mut,db="sample.fa")
res

!is.null(res)

cat("The result for any matches of the mutated sequence is",!is.null(res) )

rep_mut<-function(NMUT){
  tophit_mut<-mutator(myseq=tophit,nmut=NMUT)
  res<-myblastn_tab(myseq=tophit_mut,db="sample.fa")
  as.numeric(!is.null(res))
}

replicate(n=100, rep_mut(200))

#The sites that we have changed for our 650bp sequence.
n_mut_range<-c(0,30,50,80,110,140,170,200)

replicator_function<-function(NMUT){
  mean(replicate(n=100,rep_mut(NMUT)))
}

myresult<-sapply(n_mut_range, replicator_function)

myresult

#Daisy 
# Create function for BLAST search for a mutated sequence
matching <- function(nmut) {
  db_path <- "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa"
  mutated_seq <- mutator(myseq = tophit, nmut)
  mut_char <- seqinr::c2s(mutated_seq)
  mat_table <- myblastn_tab(mut_char, db_path)
  num_match <- dim(mat_table)[1]
  if(is.null(num_match)) {
    return(0)
  }else{
      return(1)
    }
}
probability <- c()
for (nmut in 1:length_seq) {
  prob <- mean(replicate(100, matching(nmut)))
  probability[nmut] <- prob
}
numb_mutation <- seq(1:length_seq)
prevent_search <- which(probability == 0)[1]


```
# Question 6

Now we can construct the increasing proportion of mutated bases which reduces the ability for BLAST to match the gene origin.
```{r, chart}
#To create a chart we can make a scatterplot.

#The proportion of sites randomised from the total sequence length of 650bp
x<-c(0,0.04,0.07,0.12,0.16,0.21,0.26,0.31)

  #The proportion of successful BLASTs
y<-c(1.00, 0.99, 0.70, 0.45, 0.19, 0.04)

plot(x,y, 
     main="The effect of increasing number of random bases in BLAST", 
     xlab="proportion of sites randomised", 
     ylab="proportion of succesful BLASTs")

abline(lm(y~x),col="read")
```


