---
title: "Assessment 3- Daisy Ta"
author: "Daisy Ta"
date: "5/17/2021"
output:
  html_document:
    toc: true
theme: cosmo
---

# PART 1

## Question 1 
Firstly, we use download.file to downloaded file gene_expression which contains RNA-seq count data for two samples of interest. We use read.table to read in the file and use the command header=TRUE to read files with the labels in the first row. And then, we used str to check that the structure is what they wanted. After that, we use [] to access a gene by a gene name. Finally, we output the first six rows using [1:6,].
```{r Q1}
# Download the file "gene_expression.tsv"
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/gene_expression.tsv",
              destfile= "gene_expression.tsv")
# Read in the file
gene_expression <- read.table("gene_expression.tsv", header=TRUE, row.names=1)
str(gene_expression)
# An example of trying to access a gene by gene name
gene_expression["ENSG00000227232", ]
# a table of values for the frst six genes.
gene_expression[1:6,]
```

## Question 2
Firstly, we use rowMeans to calculate the mean of the rows. Then, we make a new mean column by using $mean. Finally, we use [1:6,] to show the first six genes. 
```{r Q2}
# Calcualte the mean of the rows
row_mean <- rowMeans(gene_expression)
# Make a new mean column
gene_expression$mean <- c(row_mean)
# Show the first six genes
gene_expression[1:6,]
```
## Question 3
Firstly, we use the order command to sort the genes, the result gives us the lowest to highest mean values of the genes. Then, we use the tail command which selects the last rows of the dataset which corresponds to the top 10 highest gene expressions.
```{r Q3}
# Create sorted data frame from the ordered mean column
sorted_gene_expression <- gene_expression[order(gene_expression$mean), ]
# Get the 10 highest mean values. 
top10genes <- row.names(tail(sorted_gene_expression, 10))
top10genes
```

## Question 4
We calculate the numbers of genes with a mean < 10 to use the sum command and select means <10.We use the cat command to show the result. The number of genes with a mean < 10 is 43124.

```{r Q4}
# Calculate the numbers of genes with a mean < 10
number_genes <- sum(gene_expression$mean < 10)
number_genes
# Show the result
cat("The number of genes with a mean < 10 is", number_genes)
```

## Question 5
We use the hist function to create a histogram plot of the mean values in png format. In this case, we directly show the histogram plot by running code on Rmarkdown file. For another case, we can use the command png or dev.off to save the histogram plot. 
```{r Q5}
# Make a histogram plot of the mean values
hist(gene_expression$mean, main= "Histogram plot of the mean values", xlab= "Mean values", col="green")
```

## Question 6
We use download.file command to download the csv file and use read.csv to read in the file. Then, we use str and head command to make sure the data has been successful imported. Finally, we use colnames command to get the column names. 
```{r Q6}
# Download the file growth_data.csv
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/growth_data.csv",
              destfile= "growth_data.csv")
# Read in the file
growth_data <- read.csv("growth_data.csv", header=TRUE,stringsAsFactors = FALSE)
str(growth_data)
head(growth_data)
# Get the column names
colnames(growth_data)
```

## Question 7
We use the mean and sd command to calculate mean and standard deviation of tree circumference respectively in 2004 and 2019. Then, we use the cat command to show the results. Mean and sd of tree circumference in 2004 are 5.077 and 1.054462 respectively. Mean and sd of tree circumference in 2019 are 49.912 and 22.17979 respectively
```{r Q7}
#The mean and standard deviation of tree circumference at the start (in 2004)
mean_2004 <- mean(growth_data$Circumf_2004_cm)
cat("Mean in 2004 is", mean_2004)
sd_2004 <- sd(growth_data$Circumf_2004_cm)
cat("SD in 2004 is", sd_2004)

#The mean and standard deviation of tree circumference at the end (in 2019)
mean_2019 <- mean(growth_data$Circumf_2019_cm)
cat("Mean in 2019 is", mean_2019)
sd_2019 <- sd(growth_data$Circumf_2019_cm)
cat("SD in 2019 is", sd_2019)
```

## Question 8
Firstly, we get the tree circumference in 2004 and 2019 by using $ sign. Then, we use boxplot to make a box plot of tree circumference at the start and end of the study at both sites. The name of the box plot is "Box plot of tree circumference". 
```{r Q8}
# Get the tree circumference in 2004 and 2019
x_vals <- growth_data$Circumf_2004_cm
y_vals <- growth_data$Circumf_2019_cm
# Make a box plot
boxplot(x_vals, y_vals,
main = "Box plot of tree circumference",
names = c("2004", "2019"),
xlab = "Year",
ylab = "Tree circumference (cm)",
col = "orange"
)
```

## Question 9: Calculate the mean growth over the past 10 years at each site.
Firstly, we get the tree circumference values of each site by using subset(). Then, we use mean command to calculate the mean values in 2009 and 2019 of each site before calculate. The mean growth over the past 10 years of Northeast and Southwest are 30.076 and 48.354 respectively.
```{r Q9}
# Get the tree circumference values of each site
northeast <- subset(growth_data, Site=="northeast")
southwest <- subset(growth_data, Site=="southwest")
# Calculate the mean values of each site over past 10 years
na_2019 <- mean(northeast$Circumf_2019_cm)
sw_2019 <- mean(southwest$Circumf_2019_cm)
na_2009 <- mean(northeast$Circumf_2009_cm)
sw_2009 <- mean(southwest$Circumf_2009_cm)
# Calculate the mean growth of each site
na_mean_growth <- na_2019 - na_2009
sw_mean_growth <- sw_2019 - sw_2009
cat("The mean growth over the past 10 years of Northeast is", na_mean_growth)
cat("The mean growth over the past 10 years of Southwest is", sw_mean_growth)
```

## Question 10: Use the t.test and wilcox.test functions to estimate the p-value that the 10 year growth is different at the two sites.
We get the growth values of tree circumference on each site by using $. Then, we use t.test() and wilcox.test() to run t.test and wilcox.test before getting the p-value. We use cat() command to show the result. The p-value of t.test is 1.712524e-06 and p-value of wilcox.test is 4.6264e-06.
```{r Q10}
#get the growth values of each site
na_growth <- northeast$Circumf_2019_cm - northeast$Circumf_2009_cm
sw_growth <- southwest$Circumf_2019_cm - southwest$Circumf_2009_cm
#run t.test
t <- t.test(na_growth, sw_growth)
#get p-value of t.test
t_pvalue <- t$p.value
cat("p-value of t.test is", t_pvalue)
#run wilcox.test
wilcox <- wilcox.test(na_growth, sw_growth)
# get p-value of wilcox.test
wilcox_pvalue <- wilcox$p.value
cat("p-value of wilcox.test is", wilcox_pvalue)
```
# Part 2
```{r loading library}
# Loading the library
suppressPackageStartupMessages({
  library("seqinr")
  library("magrittr")
  library("kableExtra")
  library("R.utils")
  library("rBLAST")
  library("ape")
  library("ORFik")
  library("Biostrings")
  source("https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part2_files/mutblast_functions.R")
})
```
## Question 1
We use download.file to download the whole set of E. coli gene DNA sequences and and use R.utils::gunzip to decompress. Then, we use makeblastdb() function to create a BLAST database with dbtype is nucleic. There are 4140 sequences which present in the E.coli set. 
```{r Q1Part2}
# Download the whole set of E. coli gene DNA sequences
download.file("ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/cds/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz",
              destfile = "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz")
# Use gunzip to decompress
R.utils::gunzip("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz", overwrite=TRUE)
# Create a BLAST database
makeblastdb("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa", dbtype="nucl", "-parse_seqids")
```
## Question 2
We use download.file to download the sample fasta sequence. Then, we use seqinr::read.fasta to read them in. We use the command [] to choose our interest sequence (number 1). After that, we use getLength() and seqinr::GC() to determine the length and the proportion of GC bases respectively. Finally, we use cat() to show the results. The length of my selected sequence is 615. The proportion of GC bases is 0.5560976
```{r Q2Part2}
# Download the sample fasta sequences
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part2_files/sample.fa",
              destfile = "sample.fa")
# Read the sequences
SEQ <- seqinr::read.fasta("sample.fa")
# Select one sequence
my_seq <- SEQ[[1]]
# The length of my selected sequence
seqinr::getLength(my_seq)
# The proportion of GC bases
seqinr::GC(my_seq)
# Show the results
cat(" The length of my selected sequence is", seqinr::getLength(my_seq))
cat("The proportion of GC bases is", seqinr::GC(my_seq))
```
## Question 3
We use myblastn_tab to perform BLAST search. Then, the str command is used to see the structure. After that, we as as.character to identify what E.coli gene my sequence matches best. There is one sequence meet the demand because this gene does not share any.....  
```{r Q3Part2}
# Create BLAST databases and perform BLAST searches
myblastn_tab
res <- myblastn_tab(myseq= my_seq, db = "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa")
str(res)
#  Identify what E. coli gene my sequence matches best
top_hits <- as.character(res$sseqid[1:3])
top_hits
# Show a table of the top 3 hits
head(res, 3)[, c("qseqid", "sseqid", "pident", "evalue","bitscore")]
```
## Question 4
We use mutator() function to create mutated sequence with 100 point mutations and then compare with the original sequence by making a pairwise alignment by pairwiseAlignment from Biostrings library. Firstly, we read in my selected sequence. Then, we create a mutated copy with 100 substitutions. After that, we use DNAString() to convert to biostring. We use mmismatch() to get the number of mismatch. Finally, we show the result by using cat(). The number of mismatches between the original and mutated sequence is 80.
```{r Q4Part2}
mutator
# Read in
tophit <- seqinr::read.fasta("sample.fa")
tophit <- tophit[[1]]
str(tophit)
# Mutate
tophit_mut <- mutator(myseq = tophit, nmut = 100)
tophit_mut
## Perform pairwise alignment to prove that the mutation has worked as expected
# Convert to biostring
tophit_f <- DNAString(c2s(tophit))
tophit_mut_f <- DNAString(c2s(tophit_mut))
aln <- Biostrings::pairwiseAlignment(tophit_f,tophit_mut_f)
pid(aln)
# Get the number of mismatch
 nmismatch(aln)
# Show the result
 cat("The number of mismatches between the original and mutated sequence is", nmismatch(aln))
```
## Question 5
```{r Q5Part2}




```